app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+        if (changeNeeded) {
+          hikeFlow.value = updated
+        }
+      }
+    }
+
+    // Remove the hikes that need to be removed
+    for (hikeId in toRemove) {
+      _hikeFlowsMap.remove(hikeId)
+    }
+
+    // Add the saved hikes that were not in the list yet
+    if (_loadedHikesType == LoadedHikes.FromSaved) {
+      _savedHikesMap.forEach { (hikeId, savedHike) ->
+        if (!_hikeFlowsMap.containsKey(hikeId)) {
+          _hikeFlowsMap[hikeId] =
+              MutableStateFlow(
+                  Hike(
+                      id = savedHike.id,
+                      isSaved = true,
+                      plannedDate = savedHike.date,
+                      name = savedHike.name))
+        }
+      }
+    }
+
+    // Update the selected hike's saved status, unselect it if it's not loaded anymore
+    updateSelectedHike()
+
+    // Update the exposed list of hikes based on the map of hikes
+    updateHikeFlowsList()
+  }
+
+  private suspend fun loadSavedHikesAsync(onSuccess: () -> Unit, onFailure: () -> Unit) =
+      withContext(dispatcher) {
+        // Remember that the loaded hikes list will now only hold saved hikes
+        _loadedHikesType = LoadedHikes.FromSaved
+
+        // Update the local cache of saved hikes and add them to _hikeFlows
+        val success = refreshSavedHikesCacheAsync()
+
+        if (success) {
+          onSuccess()
+        } else {
+          onFailure()
+        }
+      }
+
+  /**
+   * Determines whether [hike] needs an update to its saved status based on [savedHike].
+   *
+   * Does not actually update the hike, that is left to the caller to do.
+   *
+   * @param hike The hike that might need an update to its saved status
+   * @param savedHike The currently known saved status of the provided hike. Null means the hike is
+   *   not saved.
+   * @return (false, [hike]) if the hike does not need an update, (true, updated) if the hike does
+   *   need an update, with updated being the hike with the new saved status.
+   */
+  private fun hikeNeedsSavedStatusUpdate(hike: Hike, savedHike: SavedHike?): Pair<Boolean, Hike> {
+    // If the hike is marked as saved but was unsaved, clear its saved state
+    if (savedHike == null) {
+      return if (hike.isSaved) {
+        Pair(true, hike.copy(isSaved = false, plannedDate = null))
+      } else {
+        Pair(false, hike)
+      }
+    }
+
+    // If the hike is saved, check if one of the attributes changed (assume the name won't change)
+    return if (!hike.isSaved || hike.plannedDate != savedHike.date) {
+      Pair(true, hike.copy(isSaved = true, plannedDate = savedHike.date))
+    } else {
+      Pair(false, hike)
+    }
+  }
+
+  private suspend fun saveHikeAsync(hikeId: String, onSuccess: () -> Unit, onFailure: () -> Unit) =
+      withContext(dispatcher) {
+        val successful: Boolean
+        _hikesMutex.withLock { successful = saveHikeAsyncInternal(hikeId) }
```

Comment:
Yes, it was because of code complexity, which rises quickly because `withContext(dispatcher)` is considered as one level of indentation.

I think it's best and easier to keep it in a separate function. It is the most straightforward way to make Sonar happy.